src/libraries/System.IO.Compression/src/System/IO/Compression/AesStream.cs

```diff
@@ -0,0 +1,316 @@
+            Buffer.BlockCopy(derivedKey, _key.Length, _hmacKey, 0, _hmacKey.Length);
+            Buffer.BlockCopy(derivedKey, _key.Length + _hmacKey.Length, _passwordVerifier, 0, _passwordVerifier.Length);
+
+            _hmac.Key = _hmacKey;
+        }
+
+        private void InitCipher()
+        {
+            if (_key is null)
+                throw new InvalidOperationException("Keys have not been generated.");
+
+            _aes.Key = _key;
+            _aesEncryptor = _aes.CreateEncryptor();
+        }
+
+        private void WriteHeader()
+        {
+            if (_headerWritten) return;
+
+            if (_salt is null || _passwordVerifier is null)
+                throw new InvalidOperationException("Keys have not been generated.");
+
+            _baseStream.Write(_salt);
+            _baseStream.Write(_passwordVerifier);
+
+            if (_ae2 && _crc32ForHeader.HasValue)
+            {
+                Span<byte> crcBytes = stackalloc byte[4];
+                BitConverter.TryWriteBytes(crcBytes, _crc32ForHeader.Value);
+                _baseStream.Write(crcBytes);
+            }
+
+            _headerWritten = true;
+        }
+
+        private void ReadHeader()
+        {
+            if (_headerRead) return;
+
+            int saltSize = _keySizeBits / 16;
+            _salt = new byte[saltSize];
+            _baseStream.ReadExactly(_salt);
+
+            byte[] verifier = new byte[2];
+            _baseStream.ReadExactly(verifier);
+
+            // WinZip AES uses SHA1 for PBKDF2
+            byte[] derivedKey = Rfc2898DeriveBytes.Pbkdf2(_password.Span, _salt, 1000, HashAlgorithmName.SHA1, (_keySizeBits / 8) + 32 + 2);
+
+            _key = new byte[_keySizeBits / 8];
+            _hmacKey = new byte[32];
+            _passwordVerifier = new byte[2];
+
+            Buffer.BlockCopy(derivedKey, 0, _key, 0, _key.Length);
+            Buffer.BlockCopy(derivedKey, _key.Length, _hmacKey, 0, _hmacKey.Length);
+            Buffer.BlockCopy(derivedKey, _key.Length + _hmacKey.Length, _passwordVerifier, 0, _passwordVerifier.Length);
+
+            if (!verifier.AsSpan().SequenceEqual(_passwordVerifier))
+                throw new InvalidDataException("Invalid password.");
+
+            _hmac.Key = _hmacKey;
+            InitCipher();
+
+            if (_ae2)
+            {
+                byte[] crcBytes = new byte[4];
+                _baseStream.ReadExactly(crcBytes);
+                // CRC can be validated later if needed
+            }
+
+            _headerRead = true;
+        }
+
+        private void ProcessBlock(byte[] buffer, int offset, int count)
+        {
+            if (_aesEncryptor is null)
+                throw new InvalidOperationException("Cipher has not been initialized.");
+
+            int processed = 0;
+            while (processed < count)
+            {
+                IncrementCounter();
+                byte[] keystream = new byte[16];
```

Comment:
This makes many small allocations over the course of the encryption. At the very least, move this out of the while loop. Consider using stackalloced memory or renting an array from an array pool.